docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+authorization and authentication.
+
+
+## Implementation Details
+
+### Thirdpartyapiservers Extensions Resource
+
+API server will support a non-namespaced `thirdpartyapiservers` extension resource.
+Third party servers can POST at
+`http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to register themselves.
+Clients can GET `http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to get a list of registered third party servers.
+The API definition will be as follows:
+
+```
+// ThirdPartyAPIServer represents the configuration of a third party server.
+type ThirdPartyAPIServer struct {
+  unversioned.TypeMeta
+
+  ObjectMeta
+
+  Spec ThirdPartyAPIServerSpec
+
+  Status ThirdPartyAPIServerStatus
+}
+
+// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
+// resource.
+type ThirdPartyAPIServerSpec struct {
+  // Base path of the API server.
+  BasePath string
+
+  // Path to the swagger spec.
+  SwaggerSpecPath string
+}
+
+type ThirdPartyAPIServerStatus struct {
+  Conditions []ThirdPartyAPIServerCondition
+}
+
+type ThirdPartyAPIServerConditionType string
+const (
+        ThirdPartyAPIServerValid ThirdPartyAPIServerConditionType = "Valid"
+)
+
+type ThirdPartyAPIServerCondition struct {
+  Type ThirdPartyAPIServerConditionType
+  // Status of the condition, one of True, False or Unknown.
+  Status ConditionStatus
+  Reason string
+  Message string
+}
+```
+
+There are a few requirements that the third party API servers should meet
+(for ex: name of resources provided by each server should be unique).
+ThirdPartyAPIServerController will look for new servers and will validate them.
+It will set the valid condition in the status accordingly.
+Note that the server is validated only once so it is possible to change it after
+the validation is done. We expect that the server will be deleted and added
+again, whenever its functionality is changed.
+
+### kubectl
+
+Users will be able to use resources provided by third party servers seamlessly
+as if they were provided by the core server.
+Kubectl today uses a RESTMapper to determine the list of valid resources and
+operations. We will switch it to use the discovery API so that we can add and remove resources
+at runtime. It will first get the list of third party servers from the core server
+and then use their discovery APIs to get the list of resources and operations
+supported by each one of them.
+
+### Events
+
+Core server will store all events. Third party servers will POST their events to
+the core server.
+As a result, users will get all events in the cluster, when they run `kubectl
+get events` or GET `kubernetes-master/api/v1/events`.
+
+### Auth
+
+Each API server will be resposible for authentication and authorisation for
+their own resources. Right now, the code will be duplicated in all servers.
+In future, we can provide a service which all of them can call.
```

Comment:
Rephrase:

* admins may choose to apply homogenous auth, proxy auth, or delegate auth.  Future proposals will refine how auth across a cluster is managed.